### ISSUE TITLE
[BUG] Uploading in a folder with no permissions causes a loop

### ISSUE THREAD

**Issue by jesmrec** [owncloud/ios-app#623]
I reproduce the problem by using the regex control from server version 10.4

To set up:

open file `config/config.php`
add the rule:

`'blacklisted_files_regex' => [
  '.*\.jpg'
  ]`

so, all the files matching the rule will be blacklisted and not allowed to be synced.

### Steps to reproduce
1.  Set up (as above)
2. Upload content matching the rule (a jpg image or whatever)

### Current behaviour
UI gets in loop with the error (probably related with https://github.com/owncloud/ios-app/pull/616)

![Feb-20-2020 14-44-13](https://user-images.githubusercontent.com/14894746/74939389-1226a380-53f0-11ea-9722-9885ecde3028.gif)

even killing the app and reopening, the loop goes on

### Expected behaviour
Error message to notify and no more warnings.

### Client
**iOS version:** 13.3

**ownCloud app version:** 1.3.0

**Device model:** iPhoneX

**App logs**: 
[ownCloud_Feb_21__2020_at_8_34_13_AM.log.txt](https://github.com/owncloud/ios-app/files/4234814/ownCloud_Feb_21__2020_at_8_34_13_AM.log.txt)

### Server configuration

**ownCloud version:** 10.4RC3

Note: also reproducible 

- By sharing a folder with no permissions with another user, and the recipient trying to upload new content to the shared folder. 
- By uploading content to a external storage in "only view" mode (server >= 10.4)

**Comment by michaelstingl** [comment]
@jesmrec could you post the full server response? (debug log or `mitmproxy`)

**Comment by jesmrec** [comment]
> @jesmrec could you post the full server response? (debug log or mitmproxy)

sure. App logs added to the first message

**Comment by michaelstingl** [comment]
This is what we get from the server:

```
# RESPONSE --------------------------------------------------------
Method:     PUT
URL:        http://10.40.40.198:28000/remote.php/dav/files/user1/Test/IMG_0025.jpg
Request-ID: 1208A71F-E843-4898-91AF-8D76B8755BDF
Error:      -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
403 FORBIDDEN
Content-Type: application/xml; charset=utf-8
Keep-Alive: timeout=5, max=100
Pragma: no-cache
Content-Security-Policy: default-src 'none';
Set-Cookie: oc4z5henty72=7nshlcm2n41vafqbi49oqva9lk; path=/; HttpOnly, oc_sessionPassphrase=Ru7xBbmAXTMATsneT9QDF86Exbhc4v%2F2JPk3pIPCwCLvB4vS4qA4lxLUwDbA2qSxh1820pf%2F1SqkZs0gTBkfM0Ckd3Sh0sPsdVp%2Bh6ofDryRwdhuxM60ZIDD4tfONwYg; path=/; HttpOnly, oc4z5henty72=fpi4fdgi7f5lt6muo5bietafil; path=/; HttpOnly, cookie_test=test; expires=Fri, 21-Feb-2020 08:34:27 GMT; Max-Age=3600
Server: Apache
X-XSS-Protection: 1; mode=block
X-Download-Options: noopen
Expires: Thu, 19 Nov 1981 08:52:00 GMT
X-Permitted-Cross-Domain-Policies: none
Cache-Control: no-store, no-cache, must-revalidate
Date: Fri, 21 Feb 2020 07:34:27 GMT
X-Robots-Tag: none
Content-Length: 182
Connection: Keep-Alive
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN

<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\Forbidden</s:exception>
  <s:message/>
</d:error>
```

It would be cool if we we could get a better response from the server.

**Comment by michaelstingl** [comment]
@jesmrec How does the response look like if the upload is blocked with `files_firewall` ?

**Comment by michaelstingl** [comment]
@mmattel any idea? /cc @pmaier1

**Comment by jesmrec** [comment]
@michaelstingl in every scenario with `403`, same behaviour. Check the notes in the bottom of the first message, with other scenarios in which it is reproducible.

**Comment by mmattel** [comment]
For me, and don‘t get me wrong, the server correctly sends a 403 (forbidden) for this case.
The IOS client on the other hand side, does not respect this answer and continues to try.
Mommy I place my shoes there - no - mommy I place... (everyone having children knows this phase)
Imho it is necessary to fix the child process.
You need to decide what you want to do with an answer which currently does not satisfy.
Referencing #616 you have to deal with `403`, `412` and whatever additionally unsatisfying return codes may come.

**Comment by michaelstingl** [comment]
I was wondering if the server could send more explanation in the body of the response. Then the clients could display a more actionable for the users.

**Comment by michaelstingl** [comment]
Related: https://github.com/owncloud/client/issues/7758

**Comment by mmattel** [comment]
@michaelstingl @HanaGemela 
The issue with having a more explanationable message text coming from the server is the following:

Imagine all the different storage types you can mount. variants of smb, variants of ftp, s3, variants of local... And the possible request types like write, update/change, delete
There has never been a unified messaging over the protocols about errors happened.
The only thing you can say for sure is:
You want to do something, you get variants of a "not possible" return code.
It is the combination (request, target, response) you have to look on.
So, either create a sophisticated centralized (core) based mechanism to check the combination and create a meaningful error message or you do something on the client side.
Give it a try, as I have written above, and check the responses with different mount types having on the file server side a read only permission. This will give you an idea about what you are looking for.

The only other thing I could think of additionally is, but I do not know if this is possible, to retrieve from core the read-only attribute of the mount setting.
Note: we know for sure if the mount setting has the attribute, but we would have to query the target by trying in case if not set...

**Comment by mmattel** [comment]
Imagine an additional case: you mount a storage normally, but on the storage side you put the permissions to read only. This is just a permission shift to the storage side, where you do not know what you get as deny response. So, this is not a matter where you put the read-only flag but a matter how the client reacts to a deny

**Comment by felix-schwarz** [comment]
@jesmrec Just to understand the issue better:
- does tapping `Cancel` stop the upload?
- what you'd like to see:
    - more user-friendly explanation
    - removal of `Retry` button

Is that correct?

**Comment by michaelstingl** [comment]
@felix-schwarz I think this can be improved when we add failing transfers to the status queue. Then users can react. For now, it's not nice, but it's okay.

**Comment by mmattel** [comment]
Idea. If you have a queue visibly separated by failing response code, it would be easier for a user to judge en-block for that items what to do instead of going thru them one by one... And you could name the response codes like „Forbidden (403)“. This at least would help the user identifying a cause. But this is as said just an idea.

**Comment by felix-schwarz** [comment]
@michaelstingl @jesmrec @mmattel I'm thinking of two changes here:
- use folder permissions to notify the user if they indicate that uploads are not possible when they select an upload action, *before* they spend any real time on them
- if an upload fails due to error 403, provide a better error message

I'd keep `Retry`, in case the upload failed due to a permission problem that the user can resolve.

I agree that the status queue will provide more ways to improve handling these. I've already started laying the groundwork for the status queue while working on https://github.com/owncloud/ios-app/issues/585, but put that work on hold to prioritize work on the planned 1.3.1 release.

**Comment by mmattel** [comment]
> use folder permissions to notify the user...

You have to take care. Do you know them? Only in case they are managed by core but (currently) not if managed on the storage side. See case: https://github.com/owncloud/ios-app/issues/623#issuecomment-590445178

**Comment by jesmrec** [comment]
> I'd keep Retry, in case the upload failed due to a permission problem that the user can resolve.

the only way for the user to resolve a `403` is selecting a new target folder. Forbidden condition does not use to change (is there any other way that i'm missing?)

About the two friendly options, user folder permissions sound promising. But the appropriate error notification is also valid. Whatever the option is chosen, should be implemented for every cause for a  `403` , not only uploads.

**Comment by felix-schwarz** [comment]
Couldn't the sharing user change permissions for f.ex. a shared folder so the upload can succeed?

**Comment by mmattel** [comment]
There is always a reason for Read-Only...

**Comment by jesmrec** [comment]
> Couldn't the sharing user change permissions for f.ex. a shared folder so the upload can succeed?

for sure.

but, if you share with no "create" permission is because you do not want new content there.

**Comment by jesmrec** [comment]
@felix-schwarz sorry i missed your message https://github.com/owncloud/ios-app/issues/623#issuecomment-590842420

> does tapping Cancel stop the upload?

`Cancel` does not stop neither the loop not the upload

>what you'd like to see:
>more user-friendly explanation
>removal of Retry buttonAbout the options

this is the open discussion

**Comment by felix-schwarz** [comment]
I just opened #632 with a couple of changes:
- actions are only available if permissions allow their use
- a better error message

**Comment by felix-schwarz** [comment]
> > does tapping Cancel stop the upload?
> 
> `Cancel` does not stop neither the loop nor the upload

Not good. Was that upload user-initiated or initiated through automatic media upload?

**Comment by mmattel** [comment]
Auto media upload has (atm, may change) no configurable target instance/path.
Because of that, the fixed path is in a subfolder of the instance users root which ich writeable by design. The only reason (atm) for a failed write is in this case reaching a quota limit.
Anything else is always user-initiated.

**Comment by jesmrec** [comment]
Fixes are ready to go in 1.3.1. Any other related issue should be addressed in a separate ticket.
